src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -119,120 +106,285 @@ EmbeddedAssemblies::install_preload_hooks ()
 	mono_install_assembly_refonly_preload_hook (open_from_bundles_refonly, nullptr);
 }
 
-int
-EmbeddedAssemblies::TypeMappingInfo_compare_key (const void *a, const void *b)
+template<typename Key, typename Entry, int (*compare)(const Key*, const Entry*), bool use_extra_size>
+const Entry*
+EmbeddedAssemblies::binary_search (const Key *key, const Entry *base, size_t nmemb, [[maybe_unused]] size_t extra_size)
 {
-	return strcmp (reinterpret_cast <const char*> (a), reinterpret_cast <const char*> (b));
-}
+	static_assert (compare != nullptr, "compare is a required template parameter");
 
-inline const char*
-EmbeddedAssemblies::find_entry_in_type_map (const char *name, uint8_t map[], TypeMapHeader& header)
-{
-	const char *e = reinterpret_cast<const char*> (bsearch (name, map, header.entry_count, header.entry_length, TypeMappingInfo_compare_key ));
-	if (e == nullptr)
+	// This comes from the app, so let's be civil
+	if (key == nullptr) {
+		log_warn (LOG_ASSEMBLY, "Key passed to binary_search must not be nullptr");
 		return nullptr;
-	return e + header.value_offset;
+	}
+
+	// This is a coding error on our part, crash!
+	if (base == nullptr) {
+		log_fatal (LOG_ASSEMBLY, "Map address not passed to binary_search");
+		exit (FATAL_EXIT_MISSING_ASSEMBLY); // TODO: use a proper error code here
+	}
+
+	constexpr size_t size = sizeof(Entry);
+	while (nmemb > 0) {
```

Comment:
`nmemb` is a `size_t`, meaning it's unsigned, so the only way this loop will terminate is when it reaches 0, and *only* if it reaches zero.

I cannot easily determine if e.g. `nmemb -= nmemb / 2 + 1` could ever result in underflow -- never mind the rest of this loop --resulting in an infinite loop.

Reply:
It will underflow only if `nmemb == 0`, but that's the loop termination condition so it won't happen.